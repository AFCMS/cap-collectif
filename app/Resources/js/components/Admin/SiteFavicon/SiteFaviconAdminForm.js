// @flow
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Field, reduxForm, type FormProps } from 'redux-form';
import { connect } from 'react-redux';
import { createFragmentContainer, graphql } from 'react-relay';
import { Button } from 'react-bootstrap';
import component from '../../Form/Field';
import type { GlobalState, MediaFromAPI } from '../../../types';
import { type SiteFaviconAdminForm_siteFavicon } from './__generated__/SiteFaviconAdminForm_siteFavicon.graphql';
import ChangeSiteFaviconMutation from '../../../mutations/ChangeSiteFaviconMutation';
import AlertForm from '../../Alert/AlertForm';
import RemoveSiteFaviconMutation from '../../../mutations/RemoveSiteFaviconMutation';

type FormValues = {|
  +media: ?MediaFromAPI,
|};

type RelayProps = {|
  +siteFavicon: SiteFaviconAdminForm_siteFavicon,
|};

type Props = {|
  ...RelayProps,
  ...FormProps,
|};

const formName = 'site-favicon-admin';

const onSubmit = async (values: FormValues) => {
  const { media } = values;

  if (media) {
    const mediaId = media.id;
    const input = {
      mediaId,
    };

    return ChangeSiteFaviconMutation.commit({ input });
  }

  return RemoveSiteFaviconMutation.commit({ input: {} });
};

const SiteFaviconAdminForm = (props: Props) => {
  const {
    siteFavicon,
    handleSubmit,
    pristine,
    invalid,
    valid,
    submitSucceeded,
    submitFailed,
    submitting,
  } = props;
  document.querySelectorAll('link[rel="icon"]').forEach((node: HTMLElement) => {
    if (node instanceof HTMLLinkElement) { // eslint-disable-line no-undef
      node.href = siteFavicon && siteFavicon.media ? siteFavicon.media.url : '/favicon-96x96.png';
    }
  });

  return (
    <form onSubmit={handleSubmit}>
      <Field
        id="proposal_media"
        name="media"
        component={component}
        type="image"
        label={
          <span>
            <FormattedMessage id="admin.fields.theme.media" />
          </span>
        }
        help={<FormattedMessage id="favicon-parameters" />}
      />
      <Button
        type="submit"
        id="site_favicon_save"
        bsStyle="primary"
        disabled={pristine || invalid || submitting}>
        <FormattedMessage id={submitting ? 'global.loading' : 'global.save'} />
      </Button>
      <AlertForm
        valid={valid}
        invalid={invalid}
        submitting={submitting}
        submitSucceeded={submitSucceeded}
        submitFailed={submitFailed}
      />
    </form>
  );
};

const mapStateToProps = (state: GlobalState, { siteFavicon }: Props) => ({
  initialValues: {
    media: siteFavicon.media,
  },
});

const form = reduxForm({
  onSubmit,
  enableReinitialize: true,
  form: formName,
})(SiteFaviconAdminForm);

const container = connect(mapStateToProps)(form);

export default createFragmentContainer(
  container,
  graphql`
    fragment SiteFaviconAdminForm_siteFavicon on SiteImage {
      id
      media {
        id
        url
        name
      }
    }
  `,
);
