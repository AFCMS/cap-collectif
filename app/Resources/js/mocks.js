// @flow
/* eslint-env jest */
import { type FormProps } from 'redux-form';
import { type IntlShape } from 'react-intl';
import type { RelayPaginationProp, RelayRefetchProp } from 'react-relay';

export const $refType: any = null;
export const $fragmentRefs: any = null;

export const intlMock: IntlShape = {
  locale: 'fr-FR',
  formats: {},
  messages: {},
  now: () => 0,
  formatHTMLMessage: message => String(message),
  formatPlural: message => String(message),
  formatNumber: message => String(message),
  formatRelative: message => String(message),
  formatTime: message => String(message),
  formatDate: message => String(message),
  formatMessage: message => String(message.id),
};

export const formMock: FormProps = {
  anyTouched: false,
  array: {
    insert: jest.fn(),
    move: jest.fn(),
    pop: jest.fn(),
    push: jest.fn(),
    remove: jest.fn(),
    removeAll: jest.fn(),
    shift: jest.fn(),
    splice: jest.fn(),
    swap: jest.fn(),
    unshift: jest.fn(),
  },
  asyncValidate: jest.fn(),
  asyncValidating: false,
  autofill: jest.fn(),
  blur: jest.fn(),
  change: jest.fn(),
  clearAsyncError: jest.fn(),
  clearSubmit: jest.fn(),
  destroy: jest.fn(),
  dirty: false,
  dispatch: jest.fn(),
  error: null,
  form: 'formName',
  handleSubmit: jest.fn(),
  initialize: jest.fn(),
  initialized: true,
  initialValues: {},
  invalid: false,
  pristine: true,
  reset: jest.fn(),
  submitting: false,
  submitFailed: false,
  submitSucceeded: false,
  touch: jest.fn(),
  untouch: jest.fn(),
  valid: true,
  warning: null,
};

export const relayPaginationMock: RelayPaginationProp = {
  environment: {
    applyMutation: jest.fn(),
    sendMutation: jest.fn(),
    lookup: jest.fn(),
    sendQuery: jest.fn(),
    subscribe: jest.fn(),
    streamQuery: jest.fn(),
    retain: jest.fn(),
    unstable_internal: {},
  },
  hasMore: () => false,
  isLoading: () => false,
  loadMore: jest.fn(),
  refetchConnection: jest.fn(),
};

export const relayRefetchMock: RelayRefetchProp = {
  environment: {
    applyMutation: jest.fn(),
    sendMutation: jest.fn(),
    lookup: jest.fn(),
    sendQuery: jest.fn(),
    subscribe: jest.fn(),
    streamQuery: jest.fn(),
    retain: jest.fn(),
    unstable_internal: {},
  },
  refetch: jest.fn(),
};
