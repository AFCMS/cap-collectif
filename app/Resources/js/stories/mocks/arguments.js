// @flow
import { author } from './users';

export const argument = {
  user: author,
  trashedStatus: '',
  createdAt: '1 mars 2016 à 00:00',
  publishedAt: '1 mars 2016 à 00:00',
  body:
    'In placeat reiciendis ut. Officiis praesentium quia minima ut tenetur officiis. Eaque fugit voluptates temporibus suscipit provident culpa culpa. Magni recusandae dolorem aut id.',
  votes: { totalCount: 50 },
};

export const argumentsData = [
  { ...argument, user: { ...argument.user, vip: true } },
  { ...argument },
  { ...argument, trashedStatus: 'INVISIBLE' },
  { ...argument },
];
