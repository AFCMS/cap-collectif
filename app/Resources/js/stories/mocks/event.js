// @flow
import { author } from './users';

export const event = {
  title: 'PHPTourDuTurfu',
  url: '#',
  image: {
    url: 'https://picsum.photos/300/400',
    alt: 'event picture',
  },
  date: {
    start: '2030-03-10 00:00:00',
    end: '2032-06-15 00:00:00',
  },
  location: {
    place: 'Tour Eiffel',
    zip: '75007',
    city: 'Paris',
  },
  category: 'Immobilier',

  user: author,
};
