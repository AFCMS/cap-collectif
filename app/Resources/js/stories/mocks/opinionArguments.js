// @flow
import { author } from './users';

export const opinionArgument = {
  user: author,
  trashedStatus: '',
  createdAt: '1 mars 2016 à 00:00',
  publishedAt: '1 mars 2016 à 00:00',
  body:
    'In placeat reiciendis ut. Officiis praesentium quia minima ut tenetur officiis. Eaque fugit voluptates temporibus suscipit provident culpa culpa. Magni recusandae dolorem aut id.',
  votes: { totalCount: 50 },
};

export const opinionArguments = [
  { ...opinionArgument, user: { ...opinionArgument.user, vip: true } },
  { ...opinionArgument },
  { ...opinionArgument, trashedStatus: 'INVISIBLE' },
  { ...opinionArgument },
];
