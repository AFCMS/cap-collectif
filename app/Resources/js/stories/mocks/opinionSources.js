// @flow
import { author } from './users';

export const opinionSource = {
  url: 'https://ui.cap-collectif.com/',
  user: author,
  createdAt: ' • 1 mars 2016',
  publishedAt: ' • 1 mars 2016',
  title: 'Source title',
  category: 'Politique',
  body:
    'In placeat reiciendis ut. Officiis praesentium quia minima ut tenetur officiis. Eaque fugit voluptates temporibus suscipit provident culpa culpa. Magni recusandae dolorem aut id.',
  votes: { totalCount: 2 },
};

export const opinionSources = [
  { ...opinionSource },
  { ...opinionSource },
  { ...opinionSource, user: { ...opinionSource.user, vip: true } },
  { ...opinionSource },
  { ...opinionSource, category: null },
];
