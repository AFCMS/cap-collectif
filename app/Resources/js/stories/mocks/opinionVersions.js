// @flow
import { author } from './users';

export const opinionVersion = {
  url: 'https://ui.cap-collectif.com/',
  title: 'Opinion version',
  user: author,
  createdAt: ' • 3 janvier 2015',
  pinned: true,
  publishedAt: ' • 3 janvier 2015',
  votes: { totalCount: 50 },
  arguments: { totalCount: 4 },
  sources: { totalCount: 0 },
  votesMitige: { totalCount: 1 },
  votesNok: { totalCount: 1 },
  votesOk: { totalCount: 2 },
};

export const opinionVersions = [
  { ...opinionVersion },
  { ...opinionVersion, user: { ...opinionVersion.user, vip: true } },
  { ...opinionVersion, votes: { totalCount: 0 } },
];
