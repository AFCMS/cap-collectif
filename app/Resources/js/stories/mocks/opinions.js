// @flow
import { author } from './users';

export const opinion = {
  url: 'https://ui.cap-collectif.com/',
  title: 'Opinion',
  user: author,
  createdAt: ' • 1 mars 2018',
  pinned: true,
  publishedAt: ' • 1 mars 2018',
  votes: { totalCount: 4 },
  versions: { totalCount: 3 },
  arguments: { totalCount: 4 },
  sources: { totalCount: 0 },
  votesMitige: { totalCount: 1 },
  votesNok: { totalCount: 1 },
  votesOk: { totalCount: 2 },
};

export const opinions = [
  { ...opinion },
  { ...opinion, user: { ...opinion.user, vip: true } },
  { ...opinion, votes: { totalCount: 0 } },
];
