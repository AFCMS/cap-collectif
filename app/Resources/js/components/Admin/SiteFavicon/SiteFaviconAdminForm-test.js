// @flow
/* eslint-env jest */
import * as React from 'react';
import { shallow } from 'enzyme';
import SiteFaviconAdminForm from './SiteFaviconAdminForm';
import { formMock } from '../../../mocks';

describe('<SiteFaviconAdminPage />', () => {
  it('renders correctly without a favicon set', () => {
    const props = {
      siteFavicon: {
        id: 'siteFaviconId',
        media: null,
      },
      ...formMock,
    };
    const wrapper = shallow(<SiteFaviconAdminForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when a favicon is set', () => {
    const props = {
      siteFavicon: {
        id: 'siteFaviconId',
        media: {
          id: 'mediaId',
          url:
            'https://www.gettyimages.com/gi-resources/images/CreativeLandingPage/HP_Sept_24_2018/CR3_GettyImages-159018836.jpg',
          name: 'Favicon',
        },
      },
      ...formMock,
    };
    const wrapper = shallow(<SiteFaviconAdminForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
