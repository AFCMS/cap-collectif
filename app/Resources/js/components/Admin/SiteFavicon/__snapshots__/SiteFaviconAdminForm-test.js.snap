// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteFaviconAdminPage /> renders correctly when a favicon is set 1`] = `
<ContainerConstructor
  anyTouched={false}
  array={
    Object {
      "insert": [MockFunction],
      "move": [MockFunction],
      "pop": [MockFunction],
      "push": [MockFunction],
      "remove": [MockFunction],
      "removeAll": [MockFunction],
      "shift": [MockFunction],
      "splice": [MockFunction],
      "swap": [MockFunction],
      "unshift": [MockFunction],
    }
  }
  asyncValidate={[MockFunction]}
  asyncValidating={false}
  autofill={[MockFunction]}
  blur={[MockFunction]}
  change={[MockFunction]}
  clearAsyncError={[MockFunction]}
  clearSubmit={[MockFunction]}
  componentRef={null}
  destroy={[MockFunction]}
  dirty={false}
  dispatch={[MockFunction]}
  error={null}
  form="formName"
  handleSubmit={[MockFunction]}
  initialValues={Object {}}
  initialize={[MockFunction]}
  initialized={true}
  invalid={false}
  pristine={true}
  reset={[MockFunction]}
  siteFavicon={
    Object {
      "id": "siteFaviconId",
      "media": Object {
        "id": "mediaId",
        "name": "Favicon",
        "url": "https://www.gettyimages.com/gi-resources/images/CreativeLandingPage/HP_Sept_24_2018/CR3_GettyImages-159018836.jpg",
      },
    }
  }
  submitFailed={false}
  submitSucceeded={false}
  submitting={false}
  touch={[MockFunction]}
  untouch={[MockFunction]}
  valid={true}
  warning={null}
/>
`;

exports[`<SiteFaviconAdminPage /> renders correctly without a favicon set 1`] = `
<ContainerConstructor
  anyTouched={false}
  array={
    Object {
      "insert": [MockFunction],
      "move": [MockFunction],
      "pop": [MockFunction],
      "push": [MockFunction],
      "remove": [MockFunction],
      "removeAll": [MockFunction],
      "shift": [MockFunction],
      "splice": [MockFunction],
      "swap": [MockFunction],
      "unshift": [MockFunction],
    }
  }
  asyncValidate={[MockFunction]}
  asyncValidating={false}
  autofill={[MockFunction]}
  blur={[MockFunction]}
  change={[MockFunction]}
  clearAsyncError={[MockFunction]}
  clearSubmit={[MockFunction]}
  componentRef={null}
  destroy={[MockFunction]}
  dirty={false}
  dispatch={[MockFunction]}
  error={null}
  form="formName"
  handleSubmit={[MockFunction]}
  initialValues={Object {}}
  initialize={[MockFunction]}
  initialized={true}
  invalid={false}
  pristine={true}
  reset={[MockFunction]}
  siteFavicon={
    Object {
      "id": "siteFaviconId",
      "media": null,
    }
  }
  submitFailed={false}
  submitSucceeded={false}
  submitting={false}
  touch={[MockFunction]}
  untouch={[MockFunction]}
  valid={true}
  warning={null}
/>
`;
